packages/blocks/src/page-block/edgeless/components/toolbar/shape/shape-element.ts

```diff
@@ -0,0 +1,319 @@
+import { sleep } from '@blocksuite/global/utils';
+import { WithDisposable } from '@blocksuite/lit';
+import {
+  css,
+  html,
+  LitElement,
+  type PropertyValues,
+  type TemplateResult,
+} from 'lit';
+import { customElement, property, query, state } from 'lit/decorators.js';
+
+import type { EdgelessTool } from '../../../../../__internal__/index.js';
+import {
+  Bound,
+  PhasorElementType,
+  ShapeStyle,
+  type ShapeType,
+  StrokeStyle,
+} from '../../../../../surface-block/index.js';
+import type { EdgelessPageBlockComponent } from '../../../edgeless-page-block.js';
+import {
+  DEFAULT_SHAPE_FILL_COLOR,
+  DEFAULT_SHAPE_STROKE_COLOR,
+} from '../../component-toolbar/change-shape-button.js';
+
+export type ShapeName = ShapeType | 'roundedRect';
+
+export interface Shape {
+  name: ShapeName;
+  svg: TemplateResult<1>;
+}
+
+interface Coord {
+  x: number;
+  y: number;
+}
+
+interface TransformMap {
+  [key: string]: {
+    x: number;
+    y: number;
+    scale: number;
+    origin: string;
+  };
+}
+
+@customElement('edgeless-shape-element')
+export class EdgelessShapeElement extends WithDisposable(LitElement) {
+  static override styles = css`
+    .shape {
+      --x: 0px;
+      --y: 0px;
+      --offset-x: 0px;
+      --offset-y: 0px;
+      --scale: 1;
+      transform: translateX(calc(var(--offset-x) + var(--x)))
+        translateY(calc(var(--y) + var(--offset-y))) scale(var(--scale));
+      height: 60px;
+      width: 60px;
+      display: flex;
+      justify-content: center;
+      align-items: center;
+      position: absolute;
+      top: 12px;
+      left: 16px;
+      transition: all 0.5s cubic-bezier(0, -0.01, 0.01, 1.01);
+    }
+    .shape.dragging {
+      transition: none;
+    }
+    .shape svg {
+      height: 100%;
+      filter: drop-shadow(0px 2px 8px rgba(0, 0, 0, 0.15));
+    }
+  `;
+
+  @property({ attribute: false })
+  shape!: Shape;
+
+  @property({ attribute: false })
+  order!: number;
+
+  @property({ attribute: false })
+  getContainerRect!: () => DOMRect;
+
+  @property({ attribute: false })
+  handleClick!: () => void;
+
+  @property({ attribute: false })
+  edgeless!: EdgelessPageBlockComponent;
+
+  @property({ attribute: false })
+  setEdgelessTool!: (edgelessTool: EdgelessTool) => void;
+
+  @query('#shape-element')
+  private shapeElement!: HTMLElement;
+
+  @query('#backup-shape-element')
+  private backupShapeElement!: HTMLElement;
+
+  private transformMap: TransformMap = {
+    z1: { x: 0, y: 5, scale: 1, origin: '50% 100%' },
+    z2: { x: -15, y: 0, scale: 0.75, origin: '20% 20%' },
+    z3: { x: 15, y: 0, scale: 0.75, origin: '80% 20%' },
+    hidden: { x: 0, y: 120, scale: 0, origin: '50% 50%' },
+  };
+
+  @state()
+  private startCoord: Coord = { x: 0, y: 0 };
+
+  @state()
+  private dragging: boolean = false;
+
+  @state()
+  private isOutside: boolean = false;
+
+  private onDragStart = (coord: Coord) => {
```

Comment:
Please add  `_` prefix for private methods and variables.

Reply:
done